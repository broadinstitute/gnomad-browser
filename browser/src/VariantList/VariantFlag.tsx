import React from 'react'

import { Badge } from '@gnomad/ui'

type Flag = {
  label: string
  level: 'info' | 'warning' | 'error' | 'success' | undefined
  formatTooltip: (input: any) => string
}

export const FLAGS_CONFIG: Record<string, Flag> = {
  lcr: {
    label: 'LCR',
    level: 'info',
    formatTooltip: () => 'Found in a low complexity region\nVariant annotation or quality dubious',
  },
  lc_lof: {
    label: 'LC pLoF',
    level: 'error',
    formatTooltip: (variant: any) =>
      `Low-confidence pLoF: ${variant.lof_filter}\nVariant annotation or quality dubious`,
  },
  lof_flag: {
    label: 'pLoF flag',
    level: 'warning',
    formatTooltip: (variant: any) =>
      `Flagged by LOFTEE: ${variant.lof_flags}\nVariant annotation or quality dubious`,
  },
  nc_transcript: {
    label: 'NC Transcript',
    level: 'error',
    formatTooltip: () => 'Non-protein-coding transcript\nVariant annotation dubious',
  },
  os_lof: {
    label: 'OS pLoF',
    level: 'info',
    formatTooltip: () =>
      'Other Splice Predicted Loss-of-Function: this variant is predicted to create or disrupt a splice site outside the canonical splice site (beta)',
  },
  mnv: {
    label: 'MNV',
    level: 'error',
    formatTooltip: () =>
      'Multi-nucleotide variant: this variant is found in phase with another variant in some individuals, altering the amino acid sequence\nVariant annotation dubious',
  },
  monoallelic: {
    label: 'Monoallelic',
    level: 'info',
    formatTooltip: () => 'All samples are homozygous alternate for the variant',
  },
  // Mitochondrial variants
  common_low_heteroplasmy: {
    label: 'Common Low Heteroplasmy',
    level: 'warning',
    formatTooltip: () =>
      'Variant is present at an overall frequency of .001 across all samples with a heteroplasmy level > 0 and < 0.50',
  },
}

type Props = {
  type: string
  variant: any
}

const VariantFlag = ({ type, variant }: Props) => {
  if (type in FLAGS_CONFIG) {
    const { label, level, formatTooltip } = FLAGS_CONFIG[type as keyof typeof FLAGS_CONFIG]
    return (
      <Badge level={level} tooltip={formatTooltip(variant)}>
        {label}
      </Badge>
    )
  }

  return null
}

export default VariantFlag
