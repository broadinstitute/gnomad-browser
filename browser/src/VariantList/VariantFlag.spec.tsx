import { describe, expect, test } from '@jest/globals'
import 'jest-styled-components'

import React from 'react'
import renderer from 'react-test-renderer'

import VariantFlag, { FLAGS_CONFIG } from './VariantFlag'

describe('Variant flag', () => {
  const realFlags = Object.keys(FLAGS_CONFIG)

  realFlags.forEach((flag) => {
    test(`correctly formats with key of '${flag}'`, () => {
      const tree = renderer.create(<VariantFlag key={flag} type={flag} variant={{}} />)
      expect(tree).toMatchSnapshot()
    })
  })

  test('correctly formats as nothing when given a non-existent variant flag', () => {
    const tree = renderer.create(
      <VariantFlag key="not-a-real-flag" type="not-a-real-flag" variant={{}} />
    )
    expect(tree).toMatchSnapshot()
  })
})
