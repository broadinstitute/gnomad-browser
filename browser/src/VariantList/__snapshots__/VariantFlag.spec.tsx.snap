// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variant flag correctly formats as nothing when given a non-existent variant flag 1`] = `null`;

exports[`Variant flag correctly formats with key of 'common_low_heteroplasmy' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: orange;
  color: #000;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Common Low Heteroplasmy
</span>
`;

exports[`Variant flag correctly formats with key of 'lc_lof' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: #DD2C00;
  color: #fff;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  LC pLoF
</span>
`;

exports[`Variant flag correctly formats with key of 'lcr' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: #424242;
  color: #fff;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  LCR
</span>
`;

exports[`Variant flag correctly formats with key of 'lof_flag' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: orange;
  color: #000;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  pLoF flag
</span>
`;

exports[`Variant flag correctly formats with key of 'mnv' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: #DD2C00;
  color: #fff;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  MNV
</span>
`;

exports[`Variant flag correctly formats with key of 'monoallelic' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: #424242;
  color: #fff;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Monoallelic
</span>
`;

exports[`Variant flag correctly formats with key of 'nc_transcript' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: #DD2C00;
  color: #fff;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  NC Transcript
</span>
`;

exports[`Variant flag correctly formats with key of 'os_lof' 1`] = `
.c0 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: #424242;
  color: #fff;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

<span
  className="c0"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  OS pLoF
</span>
`;
